### PR TITLE
refactor[codegen]: make settings into a global object

### PR DESCRIPTION
**Pull request by charles-cooper**
### How I did it
cherry-picked relevant commits from https://github.com/vyperlang/vyper/pull/3905

### Commit message
```
this commit adds a global settings object, unifying various functions
which modify global settings:
anchor_opt_level, anchor_evm_version, _set_debug
```
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/compiler/settings.py [hunk 1]
```diff
@@ -43,16 +44,60 @@ class Settings:
...
 
 
-_DEBUG = False
+_settings = None
 
 
-def _is_debug_mode():
-    global _DEBUG
-    return _DEBUG
+def get_global_settings() -> Optional[Settings]:
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/compiler/settings.py:53]: it's mimicking the behavior the current evm-version, optimization and debug flags (which are global). i'm also in general against global variables, but in the case of the compilation-global settings, i think it's ok (you never need to set settings "locally", although you might want to monkey patch settings in tests)

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/compiler/settings.py:53]: e.g. you can see here, evm version is set globally in the conftest:
https://github.com/vyperlang/vyper/blob/b43fface148b45cbd2c90b5d24f77398a63745b5/tests/conftest.py#L127-L136

**Review by harkal — APPROVED** [review]

File: vyper/evm/opcodes.py [hunk 2]
```diff
@@ -208,16 +207,9 @@
 IR_OPCODES: OpcodeMap = {**OPCODES, **PSEUDO_OPCODES}
 
 
-@contextlib.contextmanager
-def anchor_evm_version(evm_version: Optional[str] = None) -> Generator:
+def set_global_evm_version(evm_version: int) -> None:
     global active_evm_version
```
Comment (fubuloubu) [inline vyper/evm/opcodes.py:211]: Can this be combined into the global settings singleton?

**Review by fubuloubu — APPROVED** [review]
lgtm, one suggestion

File: vyper/evm/opcodes.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/evm/opcodes.py:211]: i think the change is a bit more extensive, but let me try

**Pull request merged**